File: fsharp/twelve-days/TwelveDays.fs
module TwelveDays

type Section = {
  day: string;
  gift: string;
}

let songSections = [|
  {
    day = "first";
    gift = "a Partridge in a Pear Tree";
  };
  {
    day = "second";
    gift = "two Turtle Doves";
  };
  {
    day = "third";
    gift = "three French Hens";
  };
  {
    day = "fourth";
    gift = "four Calling Birds";
  };
  {
    day = "fifth";
    gift = "five Gold Rings";
  };
  {
    day = "sixth";
    gift = "six Geese-a-Laying";
  };
  {
    day = "seventh";
    gift = "seven Swans-a-Swimming";
  };
  {
    day = "eighth";
    gift = "eight Maids-a-Milking";
  };
  {
    day = "ninth";
    gift = "nine Ladies Dancing";
  };
  {
    day = "tenth";
    gift = "ten Lords-a-Leaping";
  };
  {
    day = "eleventh";
    gift = "eleven Pipers Piping";
  };
  {
    day = "twelfth";
    gift = "twelve Drummers Drumming"
  };
|]

let createVerse (day: string) (gifts: string) = sprintf "On the %s day of Christmas my true love gave to me: %s." day gifts

let createGifts (day: int) =
  let iterateGifts gifts index =
    let gift = (Array.get songSections index).gift
    if index = 0 then gifts + ", and " + gift
    elif gifts = "" then gift
    else gifts + ", " + gift
  if day = 0 then (Array.get songSections day).gift
  else List.fold iterateGifts "" (List.rev [0 .. day])

let recite (start: int) (stop: int) =
  [start .. stop ] |> List.map (fun num ->
    let index = num - 1
    let day = (Array.get songSections index).day
    let gifts = createGifts index
    createVerse day gifts
  )

File: javascript/twelve-days/twelve-days.js
/**
 * Check out all my solutions to the Exercism JavaScript track:
 * github.com/slaymance/exercism/tree/main/javascript
 */

// Helper function
const range = (min, max) => [...Array(max).keys()].slice(min);

const GIFTS = [
  'twelve Drummers Drumming',
  'eleven Pipers Piping',
  'ten Lords-a-Leaping',
  'nine Ladies Dancing',
  'eight Maids-a-Milking',
  'seven Swans-a-Swimming',
  'six Geese-a-Laying',
  'five Gold Rings',
  'four Calling Birds',
  'three French Hens',
  'two Turtle Doves',
];

const DAYS = [
  'first', 'second', 'third', 'fourth', 'fifth', 'sixth', 'seventh', 'eighth', 'ninth', 'tenth', 'eleventh', 'twelfth'
];

const produceLine = line => `On the ${DAYS[line]} day of Christmas my true love gave to me: ${  GIFTS
  .slice(GIFTS.length - line)
  .concat(`${line === 0 ? '' : 'and '}a Partridge in a Pear Tree.`)
  .join(', ')  }\n`;

export const recite = (first, last) => range(first - 1, last || first)
  .map(produceLine)
  .join('\n');
